### PR TITLE
Improve external repo mangement

### LINKED ISSUE #510: Example of using external github repo not using code in armory

### LINKED ISSUE #516: External repo python paths
Should we add the `"external"` base path or the base directory of each imported git repo? We currently do the former, but it doesn't work the same as if it is pip installed.

### LINKED ISSUE #538: Add cache for adversarial datasets

### LINKED ISSUE #539: Document specifying a branch on external repo

### PR DESCRIPTION
**Pull request by seanpmorgan**
Closes #516 
Closes #539 
Closes #510

* Improve access pattern for external repo dirs

### CODE REVIEW

**Comment by davidslater** [comment]
Here's an example case that I think this update doesn't cover.

From my repo https://github.com/davidslater/armory-test-cases, run the `test.json` on this branch. (Note that if you run it off master, you get a different error that is fixed on this branch.)

```
(armory) davidslater@aleph-5:~/git/davidslater/armory$ python -m armory run /Users/davidslater/git/davidslater/armory-test-cases/test.json 
2020-05-14 12:52:46 aleph-5.local armory.docker.management[96308] INFO ARMORY Instance 38c07f278c created.
2020-05-14 12:52:46 aleph-5.local armory.eval.evaluator[96308] INFO Running evaluation script
2020-05-14 19:52:50 38c07f278c72 __main__[6] WARNING Outputs will be written to /armory/outputs/61958f54-b900-4f75-acca-ae3bf04761c3

2020-05-14 19:52:51 38c07f278c72 root[6] INFO Downloading external repo: davidslater/armory-test-cases

Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/conda/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/workspace/armory/scenarios/base.py", line 159, in <module>
    run_config(args.config, args.from_file)
  File "/workspace/armory/scenarios/base.py", line 123, in run_config
    scenario.evaluate(config)
  File "/workspace/armory/scenarios/base.py", line 40, in evaluate
    results = self._evaluate(config)
  File "/workspace/armory/scenarios/image_classification.py", line 30, in _evaluate
    classifier, preprocessing_fn = load_model(model_config)
  File "/workspace/armory/utils/config_loading.py", line 51, in load_model
    model_module = import_module(model_config["module"])
  File "/opt/conda/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/armory/tmp/61958f54-b900-4f75-acca-ae3bf04761c3/external/armory-test-cases/python_subdir/actual_modual/model_file.py", line 5, in <module>
    from actual_module.util import common
ModuleNotFoundError: No module named 'actual_module'

2020-05-14 12:52:51 aleph-5.local armory.eval.evaluator[96308] WARNING Shutting down container
2020-05-14 12:52:51 aleph-5.local armory.docker.management[96308] INFO Stopping instance: 38c07f278c
2020-05-14 12:52:51 aleph-5.local armory.eval.evaluator[96308] INFO Deleting tmp_dir /Users/davidslater/.armory/tmp/61958f54-b900-4f75-acca-ae3bf04761c3
2020-05-14 12:52:51 aleph-5.local armory.eval.evaluator[96308] INFO Removing output_dir /Users/davidslater/.armory/outputs/61958f54-b900-4f75-acca-ae3bf04761c3 if empty
```

**Comment by davidslater** [comment]
Here's the json file:
```
{
    "_description": "Fails on import of model module",
    "adhoc": null,
    "attack": {
        "knowledge": "white",
        "kwargs": {
            "eps": 0.2
        },
        "module": "art.attacks",
        "name": "FastGradientMethod"
    },
    "dataset": {
        "batch_size": 64,
        "framework": "numpy",
        "module": "armory.data.datasets",
        "name": "mnist"
    },
    "defense": null,
    "metric": {
        "means": true,
        "perturbation": "linf",
        "record_metric_per_sample": false,
        "task": [
            "categorical_accuracy"
        ]
    },
    "model": {
        "fit": true,
        "fit_kwargs": {
            "nb_epochs": 20
        },
        "model_kwargs": {},
        "module": "python_subdir.actual_modual.model_file",
        "name": "get_art_model",
        "weights_file": null,
        "wrapper_kwargs": {}
    },
    "scenario": {
        "kwargs": {},
        "module": "armory.scenarios.image_classification",
        "name": "ImageClassificationTask"
    },
    "sysconfig": {
        "docker_image": "twosixarmory/tf1:0.7.0-dev",
        "external_github_repo": "davidslater/armory-test-cases",
        "gpus": "all",
        "use_gpu": false
    }
}
```

File: docs/external_repos.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# External Repos
+You may want to include code from an external repository that is outside of your 
+current working directory project. This is supported through the `external_github_repo`
+field in the configuration file. After launch, the repository will be pulled into the 
+evaluations tmp folder and placed on the SYS PATH so that modules can be easily
```
Comment (davidslater) [inline docs/external_repos.md:5]: PYTHONPATH or SYSTEM PATH?

Comment (seanpmorgan) [reply docs/external_repos.md:5]: So it's technically SYS PATH: 
https://docs.python.org/3/library/sys.html#sys.path

It's initialized from PYTHONPATH though

Comment (davidslater) [reply docs/external_repos.md:5]: I just wanted to make it clear that it was the sys.path from python, and not the docker container's system path.

**Pull request merged**